Lib/imaplib.py

```diff
@@ -1279,6 +1389,251 @@ def print_log(self):
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle _pop({timeout})')
+
+        if not self._wait(timeout):
+            if __debug__ and imap.debug >= 4:
+                imap._mesg(f'idle _pop({timeout}) done')
+            return default
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle _pop({timeout}) reading')
+        imap._get_response()  # Reads line, calls _append_untagged()
+        resp = imap._idle_responses.pop(0)
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle _pop({timeout}) read {resp[0]}')
+        return resp
+
+    def burst(self, interval=0.1):
+        """Yield a burst of responses no more than 'interval' seconds apart
+
+        :param interval:    Time limit for each response after the first
+                            (The IDLE context's maximum duration is
+                            respected when waiting for the first response.)
+        :type interval:     int|float
+
+        This generator retrieves the next response along with any
+        immediately available subsequent responses (e.g. a rapid series of
+        EXPUNGE responses after a bulk delete) so they can be efficiently
+        processed as a batch instead of one at a time.
+
+        Represents responses as (type, datum) tuples, just as when
+        iterating directly on the context manager.
+
+        Example:
+
+        with imap.idle() as idler:
+            batch = list(idler.burst())
+            print(f'processing {len(batch)} responses...')
+
+        Produces no responses and returns immediately if the IDLE
+        context's maximum duration (the 'duration' argument) has elapsed.
+        Callers should plan accordingly if using this method in a loop.
+
+        Note: Windows IMAP4_stream connections will ignore the interval
+        argument, yielding endless responses and blocking indefinitely
+        for each one, because Windows select() only works on sockets.
+        It is therefore advised not to use this method with an IMAP4_stream
+        connection on Windows.
+        """
+        try:
+            yield next(self)
+        except StopIteration:
+            return
+
+        start = time.monotonic()
+
+        yield from iter(functools.partial(self._pop, interval, None), None)
+
+        if self._duration is not None:
+            elapsed = time.monotonic() - start
+            self._duration = max(self._duration - elapsed, 0)
+
+    def __next__(self):
+        imap = self._imap
+        start = time.monotonic()
+
+        typ, datum = self._pop(self._duration)
+
+        if self._duration is not None:
+            elapsed = time.monotonic() - start
+            self._duration = max(self._duration - elapsed, 0)
+
+        if not typ:
+            if __debug__ and imap.debug >= 4:
+                imap._mesg('idle iterator exhausted')
+            raise StopIteration
+
+        return typ, datum
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
```

Comment:
Can we have the `__exit__` near the `__enter__`? it's harder to mentally review it IMO.